modules/message-lane/src/lib.rs

```diff
@@ -0,0 +1,283 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Runtime module that allows sending and receiving messages using lane concept:
+//!
+//! 1) the message is sent using `send_message()` call;
+//! 2) every outbound message is assigned nonce;
+//! 3) the messages are stored in the storage;
+//! 4) external component (relay) delivers messages to bridged chain;
+//! 5) messages are processed in order (ordered by assigned nonce);
+//! 6) relay may send proof-of-receiving and proof-of-processing back to this chain.
+//!
+//! Once message is sent, its progress can be tracked by looking at module events.
+//! The assigned nonce is reported using `MessageAccepted` event. When message is
+//! accepted by the bridged chain, `MessagesDelivered` is fired. When message is
+//! processedby the bridged chain, `MessagesProcessed` by the bridged chain.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use crate::inbound_lane::{InboundLane, InboundLaneStorage};
+use crate::outbound_lane::{OutboundLane, OutboundLaneStorage};
+
+use bp_message_lane::{
+	InboundLaneData, LaneId, Message, MessageKey, MessageNonce, OnMessageReceived, OutboundLaneData,
+};
+use frame_support::{decl_event, decl_module, decl_storage, traits::Get, Parameter, StorageMap};
+use frame_system::ensure_signed;
+use sp_std::{marker::PhantomData, prelude::*};
+
+mod inbound_lane;
+mod outbound_lane;
+
+#[cfg(test)]
+mod mock;
+
+/// The module configuration trait
+pub trait Trait<I = DefaultInstance>: frame_system::Trait {
+	/// They overarching event type.
+	type Event: From<Event> + Into<<Self as frame_system::Trait>::Event>;
+	/// Message payload.
+	type Payload: Parameter;
+	/// Maximal number of messages that may be pruned during maintenance. Maintenance occurs
+	/// whenever outbound lane is updated - i.e. when new message is sent, or receival is
+	/// confirmed. The reason is that if you want to use lane, you should be ready to pay
+	/// for it.
+	type MaxMessagesToPruneAtOnce: Get<MessageNonce>;
+	/// Called when message has been received.
+	type OnMessageReceived: Default + OnMessageReceived<Self::Payload>;
```

Comment:
Why do we require require an instance (and `Default` here)? This will make it hard to plug-in another pallet as dispatch. IMHO the callback should rather be a static method.

Reply:
When I've been introducing `PruningStrategy` into ethereum bridge pallet, you've been asking to do `: Default`, because otherwise you have to specify the strategy type explicitly :) And if you have instance of `PruningStrategy`, then we may use rust type deduction mechanism to avoid that. So default isn't actually required here - could remove if desired

Reply:
`PruningStrategy` is a bit different, since it's not expected to be another substrate pallet - it was just a tiny piece of rust logic (or a bunch of numbers). Here, my understanding was, that we want to plug in say `Call::decode` dispatch pallet and that's going to be something like `pallet_call_dispatch::Module<Runtime>`, and I don't think it implements `Default`.

Reply:
I've remembered another reason when implementing this fix - sorry for not mentioning it earlier - working an another part now :/ We may use different `OnMessageReceived` impls in single pallet. One, default, from `fn receive_messages()`. And second one, custom, from `fn process_lane_messages()`. The reason behind that is that process-existing-messages version may use other scheduling algorithm. I.e. as you suggested elsewhere, we may process calls that are < 10% of block immediately and large calls only when `fn process_lane_messages()` is called. And that custom impl may need to store cumulative weight (or something similar) in-memory => hence instance. Plus, the fact that instance is required in this custom func, imo is a bit clearer than requiring a generic param.

That said, I have commit that removes instances [here](https://github.com/paritytech/parity-bridges-common/commit/50038e3af5e573ed5d8eb13fc8d009d03bf738d2). Can cherry-pick on demand. But imo better to leave as is.

Reply:
Let's see how it will look like when we plug in actual "production" modules and revisit then.